File: 03월 11일 - 스택, 큐, 덱/과제/1213.py
import sys
from collections import Counter
input = sys.stdin.readline

"""
 [팰린드롬 만들기] - 단순 구현 문제
 팰린드롬 문자열은 part1(part3의 대칭) + part2(최대 1글자) + part3(part1의 대칭)으로 이루어진다.
 1. 팰린드롬 문자열을 만들기 위해, 각 알파벳이 몇 개씩 존재하는지 리스트에 저장한다.
 2. 사전순으로 앞선 팰린드롬 수를 만들어야 하므로, 'A'부터 시작해서 part1을 문자의 개수에 맞춰 더해나간다.
 3. 만약 알파벳의 개수가 홀수인 경우, part2에 할당하고, 이미 part2에 문자가 있는 경우엔 팰린드롬을 만들 수 없다.
"""

def make_palindrome(text):
    # 각 알파벳의 수를 가지고 팰린드롬 문자열을 리턴하는 함수
    # 만들 수 없으면 "I'm Sorry Hansoo" 리턴
    part1 = ""
    part2 = ""

    # Counter(text): 각 문자가 몇개씩 들어있는지 dictionary 형태로 돌려줌
    # .items() : key와 value를 짝 지어서 돌려줌
    # sorted - 사전적으로 가장 앞서는 문자열을 만들기 위해 정렬
    alphabets = sorted(Counter(text).items())

    for key, value in alphabets:
        if value % 2 == 1:
            # 만약 가운데 글자가 있다면 더 이상 불가능
            if (len(part2) == 1):
                # 주의! 문자열에 '가 들어있기 때문에, ""로 감싸주어야 합니다.
                return "I'm Sorry Hansoo"
            # 비어있다면, 할당
            part2 = key
        
        part1 += key * (value//2)
    
    return part1 + part2 + part1[::-1]

# 입력
text = input().rstrip()

# 연산 + 출력
print(make_palindrome(text))

File: 03월 11일 - 스택, 큐, 덱/과제/1213.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

const int SIZE = 26;    // 알파벳의 수

vector<int> countAlphabet(string text) {
    // text에 있는 각 알파벳의 개수를 세어 리턴하는 함수
    vector<int> count(SIZE, 0); // A-Z까지 문자의 개수를 저장할 리스트
    for (int i = 0; i < text.size(); i++) {
        // 각 문자를 인덱스로 바꿔주기 위해 'A'의 인덱스를 빼줌
        count[text[i] - 'A']++;
    }
    return count;
}

string makePalindrome(vector<int> &count) {
    // 팰린드롬을 만들어 리턴하는 함수.
    string part1 = "", part2 = "", part3 = "";

    for (int i = 0; i < SIZE; i++) {
        // 만약 알파벳의 수가 홀수라면
        if (count[i] % 2 == 1) {
            // 이미 가운데 글자가 찼다면 더 이상 불가능
            if (part2.size() == 1) {
                return "I'm Sorry Hansoo";
            }
            // 가운데 글자가 비어있다면 할당
            part2 = (char) (i + 'A');
        }
        // 남은 수의 절반을 나누어 part1과 part3에 이어 붙인다
        for (int j = 0; j < count[i] / 2; j++) {
            part1 += (char) (i + 'A');
            part3 = (char) (i + 'A') + part3;
        }
    }

    return part1 + part2 + part3;
}

/*
 * [팰린드롬 만들기] - 단순 구현 문제
 * 팰린드롬 문자열은 part1(part3의 대칭) + part2(최대 1글자) + part3(part1의 대칭)으로 이루어진다.
 * 1. 팰린드롬 문자열을 만들기 위해, 각 알파벳이 몇 개씩 존재하는지 벡터에 저장한다.
 * 2. 사전순으로 앞선 팰린드롬 수를 만들어야 하므로, 'A'부터 시작해서 part1과 part3를 문자의 개수에 맞춰 더해나간다.
 * 3. 만약 알파벳의 개수가 홀수인 경우, part2에 할당하고, 이미 part2에 문자가 있는 경우엔 팰린드롬을 만들 수 없다.
 */


int main() {
    // 입력
    string input;
    cin >> input;

    // 연산 + 출력
    vector<int> count = countAlphabet(input);
    cout << makePalindrome(count);

    return 0;
}